src/Confluent.Kafka/Impl/LibRdKafka.cs

```diff
@@ -967,56 +967,47 @@ internal static ErrorCode committed(IntPtr rk, IntPtr partitions, IntPtr timeout
         internal static ErrorCode position(IntPtr rk, IntPtr partitions)
             => _position(rk, partitions);
 
-        /// <summary>
-        ///     Var-arg tag types, used in producev
-        /// </summary>
-        internal enum ProduceVarTag
-        {
-            End,       // va-arg sentinel
-            Topic,     // (const char *) Topic name
-            Rkt,       // (rd_kafka_topic_t *) Topic handle
-            Partition, // (int32_t) Partition
-            Value,     // (void *, size_t) Message value (payload)
-            Key,       // (void *, size_t) Message key
-            Opaque,    // (void *) Application opaque
-            MsgFlags,  // (int) RD_KAFKA_MSG_F_.. flags
-            Timestamp, // (int64_t) Milliseconds since epoch UTC
-            Header,    // (const char *, const void *, ssize_t) Message Header
-            Headers,   // (rd_kafka_headers_t *) Headers list
-        }
+        private delegate IntPtr Produceva(IntPtr rk,
+            rd_kafka_vu[] vus,
+            IntPtr size);
+        
+        private static Produceva _produceva;
 
-        private delegate ErrorCode Producev(IntPtr rk,
-                ProduceVarTag topicTag, string topic,
-                ProduceVarTag partitionTag, int partition,
-                ProduceVarTag vaTag, IntPtr val, UIntPtr len,
-                ProduceVarTag keyTag, IntPtr key, UIntPtr keylen,
-                ProduceVarTag msgflagsTag, IntPtr msgflags,
-                ProduceVarTag msg_opaqueTag, IntPtr msg_opaque,
-                ProduceVarTag timestampTag, long timestamp,
-                ProduceVarTag headersTag, IntPtr headers,
-                ProduceVarTag endTag);
-        private static Producev _producev;
         internal static ErrorCode producev(
-                IntPtr rk,
-                string topic,
-                int partition,
-                IntPtr msgflags,
-                IntPtr val, UIntPtr len,
-                IntPtr key, UIntPtr keylen,
-                long timestamp,
-                IntPtr headers,
-                IntPtr msg_opaque)
-            => _producev(rk,
-                    ProduceVarTag.Topic, topic,
-                    ProduceVarTag.Partition, partition,
-                    ProduceVarTag.Value, val, len,
-                    ProduceVarTag.Key, key, keylen,
-                    ProduceVarTag.Opaque, msg_opaque,
-                    ProduceVarTag.MsgFlags, msgflags,
-                    ProduceVarTag.Timestamp, timestamp,
-                    ProduceVarTag.Headers, headers,
-                    ProduceVarTag.End);
-
+            IntPtr rk,
+            string topic,
+            int partition,
+            IntPtr msgflags,
+            IntPtr val, UIntPtr len,
+            IntPtr key, UIntPtr keylen,
+            long timestamp,
+            IntPtr headers,
+            IntPtr msg_opaque)
+        {
+            IntPtr topicStrPtr = Marshal.StringToCoTaskMemAnsi(topic);
```

Comment:
there's a helper method `StringAsPinnedUTF8`, which is better i think (though the charset of kafka topics is limited):

https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/src/Confluent.Kafka/Impl/SafeKafkaHandle.cs#L272

Reply:
maybe wrap the whole everything in this method in the using block.